### PR TITLE
TST: Make ARM build work (not in the CI)

### LINKED ISSUE #41737: CI: Set up CI for ARM
xref #33971

This used to run in Travis, but seems like Travis now requires credits, and we can't get them reliably (see #38943).

[This page](https://developer.arm.com/solutions/infrastructure/developer-resources/ci-cd) contains CI systems that support ARM. Most of them require to set up our own runner, which I guess we don't want to do or maintain. So, the main options seem to be:
- drone.io
- CircleCI

CC: @bashtage

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] closes #41737

This seems to be working, see:

https://app.circleci.com/pipelines/github/datapythonista/pandas/11/workflows/2ca0487f-5a63-4f01-936c-e83827db714f/jobs/11

But tests are failing. I opened #41740 the the failures.

### CODE REVIEW

File: ci/setup_env.sh [hunk 1]
```diff
@@ -12,41 +12,30 @@ if [[ "$(uname)" == "Linux" && -n "$LC_ALL" ]]; then
...
-    echo "rm -rf "$MINICONDA_DIR""
-    rm -rf "$MINICONDA_DIR"
-fi
 
 echo "Install Miniconda"
-UNAME_OS=$(uname)
-if [[ "$UNAME_OS" == 'Linux' ]]; then
+DEFAULT_CONDA_URL="https://repo.continuum.io/miniconda/Miniconda3-latest"
+if [[ "$(uname -m)" == 'aarch64' ]]; then
+    CONDA_URL="https://github.com/conda-forge/miniforge/releases/download/4.8.5-1/Miniforge3-4.8.5-1-Linux-aarch64.sh"
```
Comment (fangchenli) [inline ci/setup_env.sh:19]: Could use the latest version `https://github.com/conda-forge/miniforge/releases/download/4.10.1-4/Miniforge3-4.10.1-4-Linux-aarch64.sh`

**Review by fangchenli — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
any idea why this is not showing up on the checks?

**Comment by datapythonista** [comment]
> any idea why this is not showing up on the checks?

The configuration file for CircleCI needs to be in master before it appears. I can add this in two steps. A hello world CI first that activates CircleCI, and this one later, so we can see it running before merging.

**Comment by jreback** [comment]
> > any idea why this is not showing up on the checks?
> 
> The configuration file for CircleCI needs to be in master before it appears. I can add this in two steps. A hello world CI first that activates CircleCI, and this one later, so we can see it running before merging.

sure that would be great

**Comment by lithomas1** [comment]
@datapythonista Are you sure we'll have enough credits for this? Looking at the plan, we get 2500 credits/week, which would be 250 minutes on Arm? Assuming a build takes 40 min, that would only be ~ 6 builds before we run out.

**Comment by datapythonista** [comment]
> @datapythonista Are you sure we'll have enough credits for this? Looking at the plan, we get 2500 credits/week, which would be 250 minutes on Arm? Assuming a build takes 40 min, that would only be ~ 6 builds before we run out.

Thanks for making the numbers. I didn't check in detail, but I assumed with the volume of pandas the plan credits could not be enough. But since CircleCI seems to be the only option, I think it's worth getting started, and then if we need to move to nightly builds or something, it'd be better than not having anything, as we've got now. I hope we have at least enough credits to run the tests once per day. Not as good as running on PRs, but I think that would be good enough.

**Comment by simonjayhawkins** [comment]
> I think it's worth getting started, and then if we need to move to nightly builds or something, it'd be better than not having anything, as we've got now.

nightly builds should be happening on python-wheels. but is disabled to save credits for the releases and other projects.

If we do nightly builds, shouldn't this be on the wheels repo.

**Comment by simonjayhawkins** [comment]
> nightly builds should be happening on python-wheels. but is disabled to save credits for the releases and other projects.

and because the job was rebuilding the release version and not from master https://github.com/MacPython/pandas-wheels/issues/102

**Comment by datapythonista** [comment]
Ok, since this is now green, and there is no obvious way to move forward with having ARM builds, I think the best is to remove the CircleCI configuration in this PR. We can still merge the rest of changes, and have ARM passing as of today. And if a solution is found to automatically test on ARM, whether it's at each PR, nightly, or whenever, things should be easier to add with the xfails and other changes here.

Does it sound good?

**Comment by simonjayhawkins** [comment]
I've not looked in detail, but maybe the xfails should not be strict. IIUC the latest arm build for python 3.8 and 3.9 on pandas-wheels passed

**Comment by simonjayhawkins** [comment]
> there is no obvious way to move forward with having ARM builds,

we have arm builds for conda for both mac (azure) and linux (drone). (IIUC no testing only build package for linux on drone), we have arm builds for PyPI for linux only (travis) (which run tests and fail to upload wheels if tests fail) 

from https://github.com/pandas-dev/pandas/issues/33971#issuecomment-762776064

> Also note that conda-forge has 1.2.0 packages for aarch64 as well as osx-arm64 packages. They have structurally better support; if you're on ARM64 right now you should strongly consider using conda-forge.


we need some testing on ci and consistency between pandas ci, conda and pip

**Comment by jreback** [comment]
@simonjayhawkins we don't have any arm builds 

drone doesn't work for us and azure is emulated and very slow 

yes actual wheels are built but w/o CI the only thing u can do us build the wheels and ignore the tests (not great but no option)

**Comment by simonjayhawkins** [comment]
Our basic problem is that the tests are failing on Travis in pandas-wheels on release and no nightly builds in pandas-wheels. The credits issue is part of the problem, but not having visibility of failing tests (on pandas ci and nightly builds) is probably more of an issue. If we cannot solve this issue with the limited credits, then we should also consider moving the pandas-wheels builds to CircleCI for consistency.

**Comment by jreback** [comment]
ignore the arm tests entirely in pandas-wheels

we likely cannot move the wheel builds easily

**Comment by simonjayhawkins** [comment]
> ignore the arm tests entirely in pandas-wheels

that's also an option

**Comment by simonjayhawkins** [comment]
> I've not looked in detail, but maybe the xfails should not be strict. IIUC the latest arm build for python 3.8 and 3.9 on pandas-wheels passed

if we ignore the tests in pandas-wheels, this won't be an issue.

**Comment by datapythonista** [comment]
I contacted CircleCI to check if they could be interested in sponsoring our ARM builds. Will update if I hear from them. For now looks like we're getting the 2,500 credits that @lithomas1 mentioned (not the 400,000 for open source projects for Linux builds). So, not much we can do with those.

I updated this PR so we don't add the CircleCI build, but still, it has everything else so the build passes when it can be added wherever. Only thing is that we now need to disable Circle CI, so the build doesn't fail after we remove the configuration. But only an admin of the pandas-dev organization can do it. @jreback do you mind going to https://app.circleci.com/settings/project/github/pandas-dev/pandas?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fpandas-dev%2Fpandas%2F30%2Fworkflows%2Ff9a6db17-510a-4eaf-a10c-b3a800e0a6c4%2Fjobs%2F30872 and clickig on "Stop builds" please?

**Comment by jreback** [comment]
> I contacted CircleCI to check if they could be interested in sponsoring our ARM builds. Will update if I hear from them. For now looks like we're getting the 2,500 credits that @lithomas1 mentioned (not the 400,000 for open source projects for Linux builds). So, not much we can do with those.
> 
> I updated this PR so we don't add the CircleCI build, but still, it has everything else so the build passes when it can be added wherever. Only thing is that we now need to disable Circle CI, so the build doesn't fail after we remove the configuration. But only an admin of the pandas-dev organization can do it. @jreback do you mind going to https://app.circleci.com/settings/project/github/pandas-dev/pandas?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fpandas-dev%2Fpandas%2F30%2Fworkflows%2Ff9a6db17-510a-4eaf-a10c-b3a800e0a6c4%2Fjobs%2F30872 and clickig on "Stop builds" please?

ok did that

**Comment by jreback** [comment]
thanks @datapythonista 

cc @simonjayhawkins *maybe* could backport to skip these tests on arm

**Comment by simonjayhawkins** [comment]
> cc @simonjayhawkins _maybe_ could backport to skip these tests on arm

sure. IIUC we want to update pandas-wheels, https://github.com/pandas-dev/pandas/pull/41739#issuecomment-852069720

@meeseeksdev backport 1.2.x

File: pandas/compat/__init__.py [hunk 2]
```diff
@@ -92,6 +92,18 @@ def is_platform_mac() -> bool:
     return sys.platform == "darwin"
 
 
+def is_platform_arm() -> bool:
+    """
+    Checking if he running platform use ARM architecture.
```
Comment (simonjayhawkins) [inline pandas/compat/__init__.py:97]: typo for follow-up

File: pandas/tests/indexes/interval/test_astype.py [hunk 3]
```diff
@@ -168,6 +170,7 @@ def test_subtype_integer_with_non_integer_borders(self, subtype):
         )
         tm.assert_index_equal(result, expected)
 
+    @pytest.mark.xfail(is_platform_arm(), reason="GH 41740")
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/interval/test_astype.py:173]: This one is already xfailed on 1.2.x, xref #38719

File: pandas/tests/window/test_rolling.py [hunk 4]
```diff
@@ -1072,6 +1073,7 @@ def test_rolling_sem(frame_or_series):
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.xfail(is_platform_arm(), reason="GH 41740")
```
Comment (fangchenli) [inline pandas/tests/window/test_rolling.py:1076]: This only fails on ARM64 Linux. It passes on M1 Mac. #38921

**Pull request merged**